lib/libc/sys/ioctl.c

```diff
@@ -88,9 +92,22 @@ _ioctl(int fd, unsigned long com, ...)
 		} else {
 			data = va_arg(ap, void *);
 		}
-		va_end(ap);
 	} else {
-		data = NULL;
+#ifdef __CHERI_PURE_CAPABILITY__
+		if (cheri_getlen((void*)ap) == sizeof(void *))
+			data = va_arg(ap, void *);
+		else if (cheri_getlen((void*)ap) == sizeof(uint64_t))
+			data = (void *)(intptr_t)va_arg(ap,  int64_t);
+		else if (cheri_getlen((void*)ap) == sizeof(int32_t))
+			data = (void *)(intptr_t)va_arg(ap, int32_t);
+		else
+			data = NULL;
+#else
+		/* Assume this was passed in a register. */
```

Comment:
Ideally we just wouldn't use this file at all for !purecap.  Not clear if var-args will have the same calling convention on RISC-V as on MIPS?